service/history/queues/slice.go

```diff
@@ -0,0 +1,370 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package queues
+
+import (
+	"fmt"
+
+	ctasks "go.temporal.io/server/common/tasks"
+	"go.temporal.io/server/service/history/tasks"
+)
+
+type (
+	Slice interface {
+		Scope() Scope
+		CanSplitByRange(tasks.Key) bool
+		SplitByRange(tasks.Key) (left Slice, right Slice)
+		SplitByPredicate(tasks.Predicate) (pass Slice, fail Slice)
+		CanMergeByRange(Slice) bool
+		MergeByRange(Slice) Slice
```

Comment:
`CanMergeByRange` -> `CanMergeWithSlice`

Reply:
`MergeByRange` -> `MergeWithSlice`

Reply:
i believe slice should support these 3 APIs
* SplitByRange
* SplitByPredicate
* MergeWithSlice

it would be nature to merge 2 slice and get a list of slices
merging 2 slice with same predicates require caller side pre-processing

Reply:
Yes, that's the implementation I have now. and MergeWithSlice will handle the case where predicates are the same. Plz check.